libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -290,12 +290,14 @@ void WaveClip::OnProjectTempoChange(
 
 void WaveClip::StretchLeftTo(double to)
 {
-   const auto pet = GetPlayEndTime();
-   if (to >= pet)
+   const auto sequenceDuration = GetNumSamples().as_double() * GetStretchRatio() / mRate;
+   const auto pet = mSequenceOffset + sequenceDuration;
+   to = std::min(to, pet);
+   
+   const auto ratioChange = (pet - to) / (sequenceDuration - mTrimLeft - mTrimRight);
```

Comment:
I'm having trouble with this change. Say clip looks like this:
```
0 1 2 3 4 5 6 7 // time
[   (     ) * ]
```
where `[` is sequence start time, `(` is play start time, and `*` is `to`
`ratioChange` would be `(7 - 6)/3`, does that make sense? It shouldn't even be allowed to try and stretch from the left past the play end time. I know that the border handle code prevents this, but by itself this implementation doesn't look correct.

I'm reading the commit's comment, and I understand that snapping may bring the argument `to` to be something other than the sampling period. You made me aware of this and now I can easily reproduce not nice situations, e.g. by setting the project rate to 7Hz (that's allowed) and the snapping to deciseconds. Stretching yields ugly results.

But I'm wondering if it wouldn't be simpler to for the snapping to the project sample rate period, even if the UI says otherwise. Users setting triplet snapping would I'm sure not mind a small misalignment in the display if zoomed in very close.

I don't know the snapping code, but would it be possible to make such an experiment?

Reply:
> e.g. by setting the project rate to 7Hz (that's allowed) and the snapping to deciseconds. Stretching yields ugly results.

Here's an example project, very small :
[ugly stretching.zip](https://github.com/audacity/audacity/files/12879663/ugly.stretching.zip)

Reply:
> where [ is sequence start time, ( is play start time, and * is to
ratioChange would be (7 - 6)/3, does that make sense? It shouldn't even be allowed to try and stretch from the left past the play end time. I know that the border handle code prevents this, but by itself this implementation doesn't look correct.

Considering that there was an error in `pet` evaluation this quesition is no longer relevant, right?

Reply:
> I'm reading the commit's comment, and I understand that snapping may bring the argument to to be something other than the sampling period

`GetPlayStartTime` and `GetPlayEndTime` return values that are multiples of a sample length

> Stretching yields ugly results.

Could you please be more specific?

Reply:
> I don't know the snapping code, but would it be possible to make such an experiment?

All snapping should go through `ProjectSnap::SnapTime` now, so I think that would be simple

Reply:
@vsverchinsky 

> Could you please be more specific?

You can try stretching the clip in the example project above, but here's a video: https://github.com/audacity/audacity/assets/22740106/7863cbc2-3351-4021-aebe-cfcdd11ea5a5

Reply:
> Considering that there was an error in pet evaluation this quesition is no longer relevant, right?

Probably :)

Reply:
I really do not understand that snap-to-sample strategy considering that you can easily cut-off half of a sample inside a clip if you stretch it to 50% then trim it by 1 sample.

Reply:
Visually this last sample will still look like one, but when rendered, two will be generated. By cutting-off half a sample of a 50% clip you effectively remove one complete sample from the rendering.